### PR TITLE
[git-semver-tags]: BREAKING: switch params, make it normal node-style signature

### PR DESCRIPTION
**Pull request by tunnckoCore**
e.g. `(opts, callback)`

Also, allow passing `opts.cwd` ... because it currently fails if you want to listen on another directory

Also. I would go with `Object.assign`, but was not sure for the require node version in CI.

### CODE REVIEW

**Comment by tunnckoCore** [comment]
And still. Even with `cwd` for some reason `git log` does not found that it is a git repo.

**Comment by tunnckoCore** [comment]
Updated

```js
'use strict';

import proc from 'process';
import { exec } from '@tunnckocore/execa';
import semver from 'semver';

const regex = /tag:\s*(.+?)[,)]/gi;
const cmd = 'git log --decorate --no-color';

function lernaTag(tag, pkg) {
  if (pkg && !new RegExp(`^${pkg}@`).test(tag)) {
    return false;
  }

  // eslint-disable-next-line unicorn/no-unsafe-regex
  return /^.+@\d+\.\d+\.\d+(-.+)?$/.test(tag);
}

/* eslint-disable no-param-reassign */
export default async function gitSemverTags(options) {
  const opts = Object.assign({ cwd: proc.cwd() }, options);

  if (opts.package && !opts.lernaTags) {
    throw new Error(
      'opts.package should only be used when running in lerna mode',
    );
  }

  const [{ stdout: data }] = await exec(cmd, opts);

  const tags = [];
  let tagPrefixRegexp;
  if (opts.tagPrefix) {
    tagPrefixRegexp = new RegExp(`^${opts.tagPrefix}(.*)`);
  }
  data.split('\n').forEach((decorations) => {
    let match;
    /* eslint-disable no-cond-assign */
    while ((match = regex.exec(decorations))) {
      const tag = match[1];
      if (opts.lernaTags) {
        if (lernaTag(tag, opts.package)) {
          tags.push(tag);
        }
      } else if (opts.tagPrefix) {
        const matches = tag.match(tagPrefixRegexp);
        if (matches && semver.valid(matches[1])) {
          tags.push(tag);
        }
      } else if (semver.valid(tag)) {
        tags.push(tag);
      }
    }
  });

  return tags;
}

```

let me now if you want it that way directly.

**Comment by tunnckoCore** [comment]
I'll fix the PR, there is missing `,` too.

File: packages/git-semver-tags/index.js [hunk 1]
```diff
@@ -13,7 +13,11 @@ function lernaTag (tag, pkg) {
   }
 }
 
-module.exports = function (callback, opts) {
+module.exports = function (opts, callback) {
```
Comment (evocateur) [inline packages/git-semver-tags/index.js]: I think you would need to modify all the callers of this method among the sibling packages, as well? Probably the tests, too.

Comment (tunnckoCore) [reply packages/git-semver-tags/index.js]: Oooh. Yea, sorry.

Comment (tunnckoCore) [reply packages/git-semver-tags/index.js]: Actually, tests are good. Will add one more for when there is `opts`.

**Comment by tunnckoCore** [comment]
Also. I reviewed the other packages and only two of them are using it, neither of them needs an update. They are using `gitSemverTags(cb)`, which is supported here too.

But anyway, I'll add test and update to use `Object.assign`.

**Comment by tunnckoCore** [comment]
Are you okey switching to Sindre's `execa`?

**Comment by tunnckoCore** [comment]
I don't think Promise is needed. Everyone can promisify a callback nowadays.
Also, it seems that the other packages are using callbacks too so not make much sense to turn that into a Promise API.

**Comment by bcoe** [comment]
@tunnckoCore mind rebasing? happen to be doing some work on this repo and noticed lots of stale pull requests (sorry!).

**Comment by tunnckoCore** [comment]
I'll try in one two days. (setuping the dev machine)

**Comment by bcoe** [comment]
@tunnckoCore is this still something you would like to land?

**Comment by tunnckoCore** [comment]
Uuuh, sorry. Of course.

I'll be up and running in next week.

File: packages/git-semver-tags/index.js [hunk 2]
```diff
@@ -13,36 +14,38 @@ function lernaTag (tag, pkg) {
   }
 }
 
-module.exports = function (callback, opts) {
-  opts = opts || {}
+module.exports = function gitSemverTags (opts, callback) {
```
Comment (hutson) [inline packages/git-semver-tags/index.js]: :+1: for correct callback order

File: packages/git-semver-tags/index.js [hunk 3]
```diff
@@ -13,36 +14,38 @@ function lernaTag (tag, pkg) {
...
+  if (typeof opts === 'function') {
+    callback = opts
+    opts = {}
+  }
+  var options = Object.assign({ maxBuffer: Infinity, cwd: proc.cwd() }, opts)
```
Comment (hutson) [inline packages/git-semver-tags/index.js]:
```javascript
const options = { maxBuffer: Infinity, cwd: proc.cwd(), ...opts}
```

We can use object spread now that the minimum Node version we support is Node 8 LTS, assuming we don't need to rely on the behavior of `Object.assign` to call setters (also, I changed `var` to `const`)

Comment (tunnckoCore) [reply packages/git-semver-tags/index.js]: cool cool. I should read about the diffs again, any suggestions? I started using it everywhere and still prefer it over spread for some unknown reason. ;d

**Comment by tunnckoCore** [comment]
Let me know if any other thing is needed :)

File: packages/git-semver-tags/index.js (hunk 3 above)
Comment (hutson) [reply packages/git-semver-tags/index.js]: > I should read about the diffs again, any suggestions?

I tend to just read [Mozilla documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax). :sweat_smile:  In my experience their examples are pretty good.

> I started using it everywhere and still prefer it over spread for some unknown reason. ;d

That's totally fine. My suggestion, after further consideration, was purely based on  aesthetics (from my perspective).

**Pull request merged**